src/snowflake/cli/api/utils/rendering.py

```diff
@@ -88,22 +94,151 @@ def jinja_render_from_file(
         print(rendered_result)
 
 
-class _AttrGetter:
+def _is_key_in_environ(key: str):
+    return key.upper() in os.environ or key in os.environ
+
+
+def _get_value_from_environ(key: str):
+    if key.upper() in os.environ:
+        return os.environ[key.upper()]
+    return os.environ[key]
+
+
+class _EnvGetter:
+    """
+    This class implements dot access pattern for variables that are defined as environment variables.
+    """
+
     def __init__(self, data_dict):
         self._data_dict = data_dict
 
-    def __getattr__(self, item):
+    def __getattr__(self, item: str):
+        if _is_key_in_environ(key=item):
+            return _get_value_from_environ(item)
         if item not in self._data_dict:
-            raise AttributeError(f"No attribute {item}")
+            raise AttributeError(f"No value found for {item} in environment.")
         return self._data_dict[item]
 
+    def __repr__(self):
+        return str(self._data_dict)
+
+
+class _ContextGetter:
+    """
+    Getter for project definition data. It replaces `env` attribute with _EnvGetter
+    that enables fallback to environment variables.
+    """
+
+    def __init__(self, project_definition: ProjectDefinition):
+        self._project = project_definition
+        self.variables = {v.name: v.value for v in project_definition.env}
```

Comment:
Should we not use variables, and bundle everything under env like we discussed earlier? Could the _EnvGetter handle this instead?